In [1]:
import pandas as pd
import os

In [4]:
data_path = '../data/signature_detection/signverod/'
#\data\signature_detection\signverod  ../data/signature_detection/signverod

train_df = pd.read_csv(data_path + 'train.csv')
test_df = pd.read_csv(data_path + 'test.csv')

mapping = pd.read_csv(data_path + 'image_ids.csv')

# examinando los archivos

se puede ver como los ids no coinciden con la indexación del dataset

In [5]:
mapping

height   width   id                         file_name
0     3300.0  2560.0    1                 nist_r0392_01.png
1     3300.0  2560.0    2                 nist_r0647_01.png
2     4454.0  3480.0    3      gsa_LAL50113-Lease-_Z-03.png
3     3300.0  2560.0    4                 nist_r0169_01.png
4     4411.0  3422.0    5  gsa_(R)LDC02012-Lease-SF2-04.png
...      ...     ...  ...                               ...
2760  4428.0  3445.0  628      gsa_LAZ02191-SLA-4-_Z-01.png
2761  4400.0  3400.0  629     gsa_LAZ01910-SLA-20-_Z-01.png
2762  4400.0  3400.0  630     gsa_LAR16487-SLA-05-_Z-01.png
2763  4400.0  3400.0  631      gsa_LAZ01846-SLA-3-_Z-01.png
2764  4400.0  3400.0  632      gsa_LAR17002-SLA-1-_Z-01.png

[2765 rows x 4 columns]

In [6]:
mapping.iloc[2130:2140]

height   width    id                                          file_name
2130  3300.0  2560.0  2131                                  nist_r0690_01.png
2131  3300.0  2560.0  2132                                  nist_r0883_01.png
2132  3300.0  2560.0  2133                                  nist_r0707_01.png
2133  4394.0  3456.0     1                       gsa_LAZ01910-SLA-5-_Z-01.png
2134  4400.0  3400.0     2                      gsa_LAZ02625-Lease-1_Z-05.png
2135  4414.0  3416.0     3                        gsa_LAR17002-SLA-2_Z-01.png
2136  4400.0  3400.0     4  gsa_LAR16925 - Lease Amendment - Executed (1-7...
2137  4400.0  3400.0     5                        gsa_LAL61975-Lease_Z-04.png
2138  4390.0  3424.0     6                       gsa_LAL62024-SLA-5-_Z-01.png
2139  4400.0  3400.0     7                       gsa_LAR17015-SLA-1-_Z-02.png

In [7]:
train_df

area                                               bbox  category_id  \
0     0.009  [0.14669516046674358, 0.8172596266496202, 0.33...            1   
1     0.010  [0.1830540707236842, 0.8483832465277777, 0.293...            1   
2     0.008  [0.21358420632102273, 0.8757179542824074, 0.23...            1   
3     0.006  [0.5366111738148984, 0.899244842346794, 0.0856...            2   
4     0.005  [0.6299114700902935, 0.90049396494709, 0.06553...            2   
...     ...                                                ...          ...   
7544  0.007  [0.6934562775633958, 0.12426133943781942, 0.12...            4   
7545  0.019  [0.11324846903483073, 0.6288094274776498, 0.30...            1   
7546  0.015  [0.10043294270833333, 0.7067381868657377, 0.26...            1   
7547  0.005  [0.6154690945424476, 0.09168707410562181, 0.07...            4   
7548  0.014  [0.14775272188533628, 0.7427597955706985, 0.12...            1   

        id  image_id  
0        4         2  
1        5         2  
2        6         2  
3        7         3  
4        8         3  
...    ...       ...  
7544  3281       631  
7545  3282       632  
7546  3283       632  
7547  3284       632  
7548  3285       632  

[7549 rows x 5 columns]

In [8]:
test_df

area                                               bbox  category_id  \
0     0.007  [0.1444945139800141, 0.8143371482168966, 0.284...            1   
1     0.011  [0.19762310567116137, 0.869482544646866, 0.354...            1   
2     0.007  [0.13856792273017163, 0.8392342753794001, 0.23...            1   
3     0.025  [0.4339386141765705, 0.6066057567917206, 0.308...            1   
4     0.020  [0.49798387096774194, 0.6774336998706338, 0.27...            1   
...     ...                                                ...          ...   
1661  0.001  [0.728550313313802, 0.5601104657674573, 0.0627...            3   
1662  0.001  [0.7268390909830729, 0.47641364815308873, 0.06...            3   
1663  0.010  [0.08928826649983725, 0.8310894523699259, 0.20...            1   
1664  0.020  [0.08617183685302734, 0.8941621288810808, 0.41...            1   
1665  0.006  [0.7603448869900772, 0.05906955387563884, 0.13...            4   

        id  image_id  
0        1         1  
1        2         1  
2        3         1  
3       24         8  
4       25         8  
...    ...       ...  
1661  3262       629  
1662  3263       629  
1663  3264       629  
1664  3265       629  
1665  3266       629  

[1666 rows x 5 columns]

In [9]:
train_df.iloc[4307:4316]

area                                               bbox  category_id  \
4307  0.007  [0.16437656494593575, 0.7972529619589619, 0.26...            1   
4308  0.007  [0.1630527588343576, 0.8241926171313757, 0.215...            1   
4309  0.009  [0.211285366419295, 0.8553434350024695, 0.3050...            1   
4310  0.003  [0.7267841084798177, 0.5768705938280243, 0.148...            3   
4311  0.002  [0.3158335367838542, 0.1270272756360241, 0.076...            3   
4312  0.002  [0.24785738627115886, 0.16944055458934038, 0.0...            3   
4313  0.002  [0.12151013692220052, 0.5870353069502053, 0.10...            3   
4314  0.001  [0.6939866129557292, 0.35668209901790027, 0.06...            3   
4315  0.001  [0.589381103515625, 0.35650902187701355, 0.070...            3   

        id  image_id  
4307  4735      2133  
4308  4736      2133  
4309  4737      2133  
4310     9         2  
4311    10         2  
4312    11         2  
4313    12         2  
4314    13         2  
4315    14         2

In [10]:
test_df.iloc[806:820]

area                                               bbox  category_id  \
806  0.003  [0.5694219102708804, 0.8884366819371727, 0.109...            4   
807  0.017  [0.5469214240237691, 0.31027652005174644, 0.26...            1   
808  0.023  [0.4805424528301887, 0.5741833764553687, 0.317...            1   
809  0.027  [0.5677490234375, 0.6017784233668342, 0.303911...            1   
810  0.004  [0.3234032185872396, 0.4227214022258448, 0.243...            3   
811  0.001  [0.676224365234375, 0.450071319820374, 0.07153...            3   
812  0.001  [0.6737445576985677, 0.4232982725609006, 0.072...            3   
813  0.013  [0.1152288564046224, 0.8435221193969407, 0.268...            1   
814  0.008  [0.13247891743977866, 0.792934387687623, 0.200...            1   
815  0.012  [0.1375218963623047, 0.7466988951515338, 0.271...            1   
816  0.002  [0.8187877909342448, 0.8961621382105069, 0.040...            2   
817  0.003  [0.7071158924702059, 0.10277363713798977, 0.07...            4   
818  0.002  [0.1072021484375, 0.848328260321879, 0.0632132...            2   
819  0.009  [0.31143993940509734, 0.8153066817828065, 0.11...            2   

       id  image_id  
806  4667      2100  
807  4678      2106  
808  4710      2120  
809  4716      2123  
810     1         1  
811     2         1  
812     3         1  
813     4         1  
814     5         1  
815     6         1  
816     7         1  
817     8         1  
818    54         8  
819    55         8

Para el ID de imagen necesitamos añadir 2133 a las imágenes de la segunda parte ya que la última imagen de la primera parte tiene el id 2133. Para la columna ID de la segunda parte del conjunto de datos en los archivos de prueba/entrenamiento, tenemos que añadir 4737 ya que es el mayor ID entre los dos archivos de la primera parte del conjunto de datos.

In [11]:
train_df.loc[train_df.index > 4309, 'image_id'] += 2133
train_df.loc[train_df.index > 4309, 'id'] += 4737
train_df.iloc[4307:4316]

area                                               bbox  category_id  \
4307  0.007  [0.16437656494593575, 0.7972529619589619, 0.26...            1   
4308  0.007  [0.1630527588343576, 0.8241926171313757, 0.215...            1   
4309  0.009  [0.211285366419295, 0.8553434350024695, 0.3050...            1   
4310  0.003  [0.7267841084798177, 0.5768705938280243, 0.148...            3   
4311  0.002  [0.3158335367838542, 0.1270272756360241, 0.076...            3   
4312  0.002  [0.24785738627115886, 0.16944055458934038, 0.0...            3   
4313  0.002  [0.12151013692220052, 0.5870353069502053, 0.10...            3   
4314  0.001  [0.6939866129557292, 0.35668209901790027, 0.06...            3   
4315  0.001  [0.589381103515625, 0.35650902187701355, 0.070...            3   

        id  image_id  
4307  4735      2133  
4308  4736      2133  
4309  4737      2133  
4310  4746      2135  
4311  4747      2135  
4312  4748      2135  
4313  4749      2135  
4314  4750      2135  
4315  4751      2135

In [12]:
test_df.loc[test_df.index > 809, 'image_id'] += 2133
test_df.loc[test_df.index > 809, 'id'] += 4737
test_df.iloc[806:820]

area                                               bbox  category_id  \
806  0.003  [0.5694219102708804, 0.8884366819371727, 0.109...            4   
807  0.017  [0.5469214240237691, 0.31027652005174644, 0.26...            1   
808  0.023  [0.4805424528301887, 0.5741833764553687, 0.317...            1   
809  0.027  [0.5677490234375, 0.6017784233668342, 0.303911...            1   
810  0.004  [0.3234032185872396, 0.4227214022258448, 0.243...            3   
811  0.001  [0.676224365234375, 0.450071319820374, 0.07153...            3   
812  0.001  [0.6737445576985677, 0.4232982725609006, 0.072...            3   
813  0.013  [0.1152288564046224, 0.8435221193969407, 0.268...            1   
814  0.008  [0.13247891743977866, 0.792934387687623, 0.200...            1   
815  0.012  [0.1375218963623047, 0.7466988951515338, 0.271...            1   
816  0.002  [0.8187877909342448, 0.8961621382105069, 0.040...            2   
817  0.003  [0.7071158924702059, 0.10277363713798977, 0.07...            4   
818  0.002  [0.1072021484375, 0.848328260321879, 0.0632132...            2   
819  0.009  [0.31143993940509734, 0.8153066817828065, 0.11...            2   

       id  image_id  
806  4667      2100  
807  4678      2106  
808  4710      2120  
809  4716      2123  
810  4738      2134  
811  4739      2134  
812  4740      2134  
813  4741      2134  
814  4742      2134  
815  4743      2134  
816  4744      2134  
817  4745      2134  
818  4791      2141  
819  4792      2141

In [13]:
mapping.loc[mapping.index > 2132, 'id'] += 2133
mapping.iloc[2130:2140]

height   width    id                                          file_name
2130  3300.0  2560.0  2131                                  nist_r0690_01.png
2131  3300.0  2560.0  2132                                  nist_r0883_01.png
2132  3300.0  2560.0  2133                                  nist_r0707_01.png
2133  4394.0  3456.0  2134                       gsa_LAZ01910-SLA-5-_Z-01.png
2134  4400.0  3400.0  2135                      gsa_LAZ02625-Lease-1_Z-05.png
2135  4414.0  3416.0  2136                        gsa_LAR17002-SLA-2_Z-01.png
2136  4400.0  3400.0  2137  gsa_LAR16925 - Lease Amendment - Executed (1-7...
2137  4400.0  3400.0  2138                        gsa_LAL61975-Lease_Z-04.png
2138  4390.0  3424.0  2139                       gsa_LAL62024-SLA-5-_Z-01.png
2139  4400.0  3400.0  2140                       gsa_LAR17015-SLA-1-_Z-02.png

Parece que hemos solucionado el problema. He visto que puede haber otro problema de fuga de datos de entrenamiento (datos duplicados en entrenamiento/prueba). Vamos a comprobar si existe y tratar de solucionarlo

In [14]:
testIDS = set(test_df['id'])
trainIDS = set(train_df['id'])
duplicated = testIDS.intersection(trainIDS)
len(duplicated)

1193

In [15]:
min(duplicated)

26

In [16]:
train_df.loc[train_df['id'] == 26, :]

area                                               bbox  category_id  id  \
20  0.009  [0.12200389583372734, 0.791304347826087, 0.262...            1  26   

    image_id  
20         9

In [17]:
test_df.loc[test_df['id'] == 26, :]

area                                               bbox  category_id  id  \
5  0.009  [0.12200389583372734, 0.791304347826087, 0.262...            1  26   

   image_id  
5         9

Efectivamente, aquí hay duplicación de datos.Veo dos maneras de resolver esto: eliminar algunos datos del conjunto de entrenamiento/prueba, o fusionar dos archivos sin duplicación para que uno pueda reutilizarlos y hacer una división a su elección.Yo me quedaría con la segunda opción. Debería haber:

len(test) + len(train) - len(duplicated) = 1666 + 7549 - 1193 = 8022 total rows

In [18]:
data = pd.concat([train_df, test_df]).drop_duplicates().sort_values(['id'])

In [19]:
data

area                                               bbox  category_id  \
0     0.007  [0.1444945139800141, 0.8143371482168966, 0.284...            1   
1     0.011  [0.19762310567116137, 0.869482544646866, 0.354...            1   
2     0.007  [0.13856792273017163, 0.8392342753794001, 0.23...            1   
0     0.009  [0.14669516046674358, 0.8172596266496202, 0.33...            1   
1     0.010  [0.1830540707236842, 0.8483832465277777, 0.293...            1   
...     ...                                                ...          ...   
7544  0.007  [0.6934562775633958, 0.12426133943781942, 0.12...            4   
7545  0.019  [0.11324846903483073, 0.6288094274776498, 0.30...            1   
7546  0.015  [0.10043294270833333, 0.7067381868657377, 0.26...            1   
7547  0.005  [0.6154690945424476, 0.09168707410562181, 0.07...            4   
7548  0.014  [0.14775272188533628, 0.7427597955706985, 0.12...            1   

        id  image_id  
0        1         1  
1        2         1  
2        3         1  
0        4         2  
1        5         2  
...    ...       ...  
7544  8018      2764  
7545  8019      2765  
7546  8020      2765  
7547  8021      2765  
7548  8022      2765  

[8022 rows x 5 columns]

## ahora guardamos los archivos corregidos

In [20]:
data.to_csv(data_path + "full_data.csv", index=False)
mapping.to_csv(data_path + "updated_image_ids.csv", index=False)

: 